turtlesim/tutorials/teleop_turtle_key.cpp

```diff
@@ -343,23 +283,101 @@ int TeleopTurtle::keyLoop()
         break;
       case KEYCODE_Q:
         RCLCPP_DEBUG(nh_->get_logger(), "quit");
-        return 0;
+        running = false;
+        break;
+      default:
+        // This can happen if the read returned when there was no data, or
+        // another key was pressed.  In these cases, just silently ignore the
+        // press.
+        break;
+      }
+
+      if (running && (linear != 0.0 || angular != 0.0))
+      {
+        geometry_msgs::msg::Twist twist;
+        twist.angular.z = nh_->get_parameter("scale_angular").as_double() * angular;
+        twist.linear.x = nh_->get_parameter("scale_linear").as_double() * linear;
+        twist_pub_->publish(twist);
+      }
     }
-   
 
-    geometry_msgs::msg::Twist twist;
-    twist.angular.z = a_scale_*angular_;
-    twist.linear.x = l_scale_*linear_;
-    if(dirty ==true)
+    return 0;
+  }
+
+private:
+  void spin()
+  {
+    rclcpp::spin(nh_);
```

Comment:
Is the change to `spin()` the only thing that reduced the CPU usage, or is the refactor part of it too?

Reply:
On Linux and macOS, the change to `spin` is the only thing that reduced the CPU usage.  On Windows, part of the savings was indeed changing to `spin`.  But the `KeyboardReader` used to be spinning in a tight loop looking for keyboard input, which would use up a ton of CPU time.  That is now changed to do something similar to Linux, where it blocks for 100 milliseconds looking for a keyboard input.  If it gets a key in that time, it returns it.  If not, it returns 0 so that the upper layers can do other things (like checking if we got a signal).